### PR TITLE
New marshall olkin copula

### PR DESCRIPTION
**Pull request by mbaudin47**
Creates a new MarshallOlkin copula.

### CODE REVIEW

**Comment by mbaudin47** [comment]
Good to review.

File: lib/src/Uncertainty/Distribution/openturns/MarshallOlkinCopula.hxx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  : public CopulaImplementation
+{
+  CLASSNAME
+public:
+
+  /** Default constructor */
+  MarshallOlkinCopula();
+
+  /** Parameters constructor */
+  explicit MarshallOlkinCopula(const Scalar alpha,
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/openturns/MarshallOlkinCopula.hxx]: no need for explicit keyword here : more than 1 required parameter

File: python/src/MarshallOlkinCopula_doc.i.in [hunk 2]
```diff
@@ -0,0 +1,78 @@
+%feature("docstring") OT::MarshallOlkinCopula
+"MarshallOlkin copula.
+
+Available constructor:
+    MarshallOlkinCopula(*alpha=0.0,beta=0.0*)
```
Comment (jschueller) [inline python/src/MarshallOlkinCopula_doc.i.in]: No need to list the available ctors when there's only one.

File: python/src/MarshallOlkinCopula_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+Parameters
+----------
+theta : float, :math:`-1 \leq \theta \leq 1`
+    Parameter :math:`\theta`.
```
Comment (jschueller) [inline python/src/MarshallOlkinCopula_doc.i.in]: alpha and beta ?

Comment (mbaudin47) [reply python/src/MarshallOlkinCopula_doc.i.in]: You are right: i'll fix this.

File: python/src/MarshallOlkinCopula_doc.i.in (hunk 2 above)
Comment (mbaudin47) [reply python/src/MarshallOlkinCopula_doc.i.in]: Ok.

File: lib/src/Uncertainty/Distribution/openturns/MarshallOlkinCopula.hxx (hunk 1 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/openturns/MarshallOlkinCopula.hxx]: I did not know. I'll fix this.

**Comment by sofianehaddad** [comment]
@mbaudin47 could you please squash the commits and mention the new class in `ChangeLog`?

**Comment by mbaudin47** [comment]
> @mbaudin47 could you please squash the commits and mention the new class in `ChangeLog`?

Thank you for the review. This is done.

**Comment by sofianehaddad** [comment]
@mbaudin47 please fix conflicting files

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(MarshallOlkinCopula)
+
+static const Factory<MarshallOlkinCopula> Factory_MarshallOlkinCopula;
+
+/* Default constructor */
+MarshallOlkinCopula::MarshallOlkinCopula()
+  : CopulaImplementation()
+  , alpha_(0.)
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Don't use 0 as a default value for alpha and beta as it is an invalid value for both arguments according to setAlpha() and setBeta(). A value of 0.5 is probably better.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+{
+  return ((alpha_ == 1.0) && (beta_ == 1.0));
+}
+
+/* Alpha accessor */
+void MarshallOlkinCopula::setAlpha(const Scalar alpha)
+{
+  if ((alpha <= 0.0) || (alpha >= 1.0)) throw InvalidArgumentException(HERE) << "Alpha MUST be in (0, 1), here alpha=" << alpha;
+  alpha_ = alpha;
+}
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: You must add `isAlreadyComputedCovariance_ = false;` as a change in alpha has a consequence on the covariance. Would it be possible to add 0 and 1 as limiting cases for this copula?

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+Scalar MarshallOlkinCopula::getAlpha() const
+{
+  return alpha_;
+}
+
+/* Beta accessor */
+void MarshallOlkinCopula::setBeta(const Scalar beta)
+{
+  if ((beta <= 0.0) || (beta >= 1.0)) throw InvalidArgumentException(HERE) << "Beta MUST be in (0, 1), here beta=" << beta;
+  beta_ = beta;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: You must add `isAlreadyComputedCovariance_ = false;` as a change in beta has a consequence on the covariance. Would it be possible to add 0 and 1 as limiting cases for this copula?

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    lambda2 = lambda12 / beta - lambda12
+    and set lambda12 to 1.
+    */
+    const Scalar lambda12 = 1.0;
+    const Scalar lambda1 = lambda12 /alpha_ - lambda12;
+    const Scalar lambda2 = lambda12 /beta_ - lambda12;
+    const Scalar x = std::min(-logr/lambda1,-logt);
+    const Scalar y = std::min(-logs/lambda2,-logt);
+    const Scalar u = std::exp(-(lambda1+lambda12)*x);
+    const Scalar v = std::exp(-(lambda2+lambda12)*y);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Add some spaces to the four formulas

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+Scalar MarshallOlkinCopula::getBeta() const
+{
+  return beta_;
+}
+
+/* Get the Kendall concordance of the distribution */
+CorrelationMatrix MarshallOlkinCopula::getKendallTau() const
+{
+  CorrelationMatrix tauKendall(2);
+  tauKendall(1, 0) = alpha_ * beta_ / (alpha_ + beta_ - alpha_ * beta_);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: If alpha= or beta=0 is allowed, tau=0, and if alpha=beta=1, tau=1

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  return CorrelationMatrix(tauKendall);
+}
+
+/* Get the Spearman correlation of the distribution
+*/
+CorrelationMatrix MarshallOlkinCopula::getSpearmanCorrelation() const
+{
+  CorrelationMatrix rho(2);
+  rho(1, 0) = 3 * alpha_ * beta_ / (2 * alpha_ + 2 * beta_ - alpha_ * beta_);
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: If alpha= or beta=0 is allowed, rho=0, and if alpha=beta=1, rho=1

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+{
+  if (parameter.getSize() != 2) throw InvalidArgumentException(HERE) << "Error: expected 2 value, got " << parameter.getSize();
+  setAlpha(parameter[0]);
+  setBeta(parameter[1]);
+}
+
+/* Tell if the distribution has independent copula */
+Bool MarshallOlkinCopula::hasIndependentCopula() const
+{
+  return ((alpha_ == 1.0) && (beta_ == 1.0));
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: alpha==1 && beta==1 -> the copula is equal to the min copula
alpha==0 || beta==0 -> the copula is equal to the independent copula (to be double checked)

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+Uses the algorithm from Devroye (1987) and 
+presented in 
+“An Introduction to Copulas”, 2nd Edition, Nelsen, Springer. 
+Section Exercice 3.4, page 58. 
+ */
+Point MarshallOlkinCopula::getRealization() const
+{
+  Point realization(2);
+  // We use the general algorithm based on conditional CDF inversion
+  if ((alpha_==0.0) && (beta_==0.0))
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: alpha_ == 0 || beta_ == 0

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+  {
+    const Scalar u = RandomGenerator::Generate();
+    const Scalar v = RandomGenerator::Generate();
+    realization[0] = u;
+    realization[1] = v;
+  } else {
+    const Scalar r = RandomGenerator::Generate();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Add a special case for alpha==1 && beta==1:
```
{
  const Scalar u = RandomGenerator::Generate();
  realization[0] = u;
  realization[1] = u;
}
```

**Comment by mbaudin47** [comment]
The update solves the conflict, but I still have to take the review into account.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 7 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: OK

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 8 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Ok : added and tested.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 6 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Ok. Done.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 5 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Ok. Done.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 9 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Ok. Done.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 10 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Oups. That was terribly wrong! Checked, fixed and tested.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 11 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: True: that was terribly wrong. Fixed and tested.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 12 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Done.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 4 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Ok. Fixed.

**Comment by regislebrun** [comment]
+1

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+/* Get the CDF of the distribution */
+Scalar MarshallOlkinCopula::computeCDF(const Point & point) const
+{
+  const Scalar u = point[0];
+  const Scalar v = point[1];
+  const Scalar ualpha = std::pow(u, alpha_);
+  const Scalar vbeta = std::pow(v, beta_);
+  if (ualpha>vbeta)
+  {
+    const Scalar p = std::pow(u, 1-alpha_) * v;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Add a space around "-"

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    return p;
+  }
+  else
+  {
+    const Scalar p = u * std::pow(v, 1-beta_);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: idem

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+/* Get the Kendall concordance of the distribution */
+CorrelationMatrix MarshallOlkinCopula::getKendallTau() const
+{
+  CorrelationMatrix tauKendall(2);
+  if ((alpha_==0.0) || (beta_==0.0))
+  {
+    tauKendall(1, 0) = 0.0;
+  } else if ((alpha_==1.0) && (beta_==1.0)) {
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: No opening or closing brace at opening new line or line end

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+
+/* Get the Spearman correlation of the distribution
+*/
+CorrelationMatrix MarshallOlkinCopula::getSpearmanCorrelation() const
+{
+  CorrelationMatrix rho(2);
+  if ((alpha_==0.0) || (beta_==0.0))
+  {
+    rho(1, 0) = 0.0;
+  } else if ((alpha_==1.0) && (beta_==1.0)) {
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: Idem

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+Bool MarshallOlkinCopula::hasIndependentCopula() const
+{
+  return ((alpha_ == 0.0) || (beta_ == 0.0));
+}
+
+/* Alpha accessor */
+void MarshallOlkinCopula::setAlpha(const Scalar alpha)
+{
+  if ((alpha < 0.0) || (alpha > 1.0)) throw InvalidArgumentException(HERE) << "Alpha MUST be in [0, 1], here alpha=" << alpha;
+  if ((alpha == 0.0) && (beta_ == 0.0)) throw InvalidArgumentException(HERE) << "Beta="<<beta_<<" so that alpha cannot be zero";
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: The case alpha=beta=0 is OK.

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 13 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: ok

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 14 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: ok

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 15 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: ok

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 16 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: ok

File: lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx (hunk 17 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/MarshallOlkinCopula.cxx]: ok

**Pull request merged**